### PR TITLE
rgw_sal_motr: [CORTX-27630] fix multipart abort

### PR DESCRIPTION
**Pull request by DPG17**
# Change Summary

Problem: 
Multipart Abort API fails during part deletion due to invalid state of of part's metadata (Motr oid, pv id and layout id) available in memory. The invalid state is caused due to no assignment of fetched metadata to the correct in-memory data structure.

Resolution:
  After fetching part's metadata from Store, code changes are done to assign part metadata to the correct in-memory data structure, so that Motr's Open and then Delete operations on part are successful. This results in successful Abort API.

## Parent Story
_[https://jts.seagate.com/browse/CORTX-27630]:_
_Multipart Abort API (backend part) support in RGW_

## Sub Tasks covered in this PR
NA

## Exclusions in Implementation :
NA.

## Implementation/Current Flow :
During Multipart Abort, code is added to check if multipart upload id exists in storage. If no, **no_such_upload** error is returned. Also, modified code to assign loaded object part's meta to correct internal data structure, so that open and then delete Motr object operations succeed.

## Testing Done : 

<details>
  <summary> **Case 1 : Abort multipart upload (existing upload id with only part2** </summary>
Created multipart upload with only part2 as part upload. When abort api is called, abort is successful, returning no textual response. HTTP response 204 is seen in the radosgw log.
</details>

<details>
  <summary> **Case 2 : Abort multipart upload (existing upload id with no parts** </summary>
Created multipart upload with no parts uploaded. When abort api is called, abort is successful, returning no textual response. HTTP response 204 is seen in the radosgw log.
</details>

<details>
  <summary> **Case 3 : Abort multipart (provide non-existing multipart upload id)** </summary>
  Created multipart upload with no parts. When abort api is called with invalid upload id, abort is unsuccessful, returned "NoSuchUpload" error.
</details>

<details>
  <summary> **Case 4 : Complete multipart and then list multipart uploads** </summary>
 Complete multipart upload is success. When multipart upload list is invoked, the response listed nothing (as expected).
</details>

<details>
  <summary> **Case 5 : Create multipart, Upload the second part and then list parts** </summary>
 The response listed part 2 (as expected).
</details>

Signed-off-by: Dattaprasad Govekar <dattaprasad.govekar@seagate.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -2423,6 +2476,9 @@ int MotrMultipartUpload::delete_parts(const DoutPrefixProvider *dpp)
       std::unique_ptr<rgw::sal::Object> obj;
       obj = this->bucket->get_object(rgw_obj_key(part_obj_name));
       std::unique_ptr<rgw::sal::MotrObject> mobj(static_cast<rgw::sal::MotrObject *>(obj.release()));
+      // As this is part of multipart object, set category to MultiMeta
+      mobj->set_category(RGWObjCategory::MultiMeta);
+      mobj->meta = mmpart->meta;
```
Comment (DPG17) [inline src/rgw/rgw_sal_motr.cc:2481]: This code (setting meta of object) was missing earlier. Due to this, when "open_mobj()" was called, it used to read this meta from bucket index, rather than reading it from object's part index. There are no entries for object in bucket index when object is still in multipart session (i.e. Multipart Complete is not called).

**Review by bkirunge7 — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -2457,7 +2459,8 @@ int MotrMultipartUpload::delete_parts(const DoutPrefixProvider *dpp)
   do {
     rc = this->list_parts(dpp, store->ctx(), max_parts, marker, &marker, &truncated);
     if (rc == -ENOENT) {
-      rc = -ERR_NO_SUCH_UPLOAD;
+      truncated = false;
+      rc = 0;
```
Comment (DPG17) [inline src/rgw/rgw_sal_motr.cc:2463]: The rc is masked (reset to 0) here to allow aborting an empty multipart upload session.

**Review by SeagateChaDeepak — APPROVED** [review]
LGTM +1

**Comment by andriytk** [comment]
Sorry, I cannot get the problem from the above description. Could you please elaborate it?

**Comment by DPG17** [comment]
> Sorry, I cannot get the problem from the above description. Could you please elaborate it?

The Abort API was failing earlier. The reason was part's metadata (i.e. Motr oid layout id and pv id) was not assigned to mobj->meta. Due to this no assignment, the code later tried to open object which referred bucket index to fetch the details like  Motr oid, pv id and layout id. As there is no entry in bucket index, there was no Motr oid, layout id and pv id of part object. This resulted into Abort API failure. By having the assignment (at line 2481...mobj->meta = mmpart->meta;), this issue in Abort got resolved. Also some minor changes did in part listing API.

**Comment by andriytk** [comment]
Please, add this explanation above to the PR description. Thanks.

**Comment by andriytk** [comment]
@siningwuseagate, could you please review it?

File: src/rgw/rgw_op.cc [hunk 3]
```diff
@@ -6578,6 +6578,11 @@ void RGWListMultipart::execute(optional_yield y)
 
   rgw::sal::Attrs attrs;
   op_ret = upload->get_info(this, s->yield, s->obj_ctx, &placement, &attrs);
+  if (op_ret < 0) {
+    ldpp_dout(this, 0) << "Multipart part listing: Failed to get multipart info. rc=" << op_ret << dendl;
```
Comment (andriytk) [inline src/rgw/rgw_op.cc:6582]: Use `__func__` (see examples in other places).

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:2485]: Good catch!

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -2650,16 +2662,22 @@ int MotrMultipartUpload::list_parts(const DoutPrefixProvider *dpp, CephContext *
 
     ldpp_dout(dpp, 20) << "MotrMultipartUpload::list_parts(): part_num=" << info.num
                                              << " part_size=" << info.size << dendl;
-    if (info.num != expected_next)
+    ldpp_dout(dpp, 20) << "MotrMultipartUpload::list_parts() meta:oid=[" << meta.oid.u_hi << "," << meta.oid.u_lo
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:2665]: Again, use `__func__`.

File: src/rgw/rgw_sal_motr.h [hunk 5]
```diff
@@ -806,7 +806,7 @@ class MotrMultipartPart : public MultipartPart {
 
 public:
   MotrObject::Meta  meta;
-
+  MotrMultipartPart() {}
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.h:809]: Why is this needed?

File: src/rgw/rgw_op.cc (hunk 3 above)
Comment (DPG17) [reply src/rgw/rgw_op.cc:6582]: > Use `__func__` (see examples in other places).

ok

File: src/rgw/rgw_sal_motr.cc (hunk 4 above)
Comment (DPG17) [reply src/rgw/rgw_sal_motr.cc:2665]: ok

File: src/rgw/rgw_sal_motr.h (hunk 5 above)
Comment (DPG17) [reply src/rgw/rgw_sal_motr.h:809]: Removed.

File: src/rgw/rgw_op.cc (hunk 3 above)
Comment (DPG17) [reply src/rgw/rgw_op.cc:6582]: Also, I'll discard changes from this source code file, as it is Ceph common code file. Will do changes in sal motr layer to return -ERR_NO_SUCH_UPLOAD.

**Comment by kupranay** [comment]
@andriytk ,
@DPG17  have addressed the comment. I am merging it assuming it will be approved by you and merging it. Feel free to add more comments if required and @DPG17  will address it in new PR.
